File: Controllers/VacanciesController.fs
namespace TalentSync.Controllers

open System
open System.Net.Http
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Newtonsoft.Json
open TalentSync.Models

[<ApiController>]
[<Route("api/[controller]")>]
type VacanciesController (httpClientFactory: IHttpClientFactory) =
    inherit ControllerBase()

    let deserializeVacanciesList (json: string) : VacancyList =
        JsonConvert.DeserializeObject<VacancyList>(json)

    let deserializeFullVacancy (json: string) : VacancyFull =
        JsonConvert.DeserializeObject<VacancyFull>(json)

    let serialize (data: obj) : string =
        JsonConvert.SerializeObject(data)

    [<HttpGet>]
    member this.Get([<FromQuery>] query: string) : Task<IActionResult> =
        task {
            if String.IsNullOrWhiteSpace(query) then
                return this.BadRequest("query is required") :> IActionResult
            else
                let client = httpClientFactory.CreateClient()
                client.DefaultRequestHeaders.UserAgent.ParseAdd("TalentSyncApp/1.0")

                // 1 получаем список вакансий
                let listUrl = $"https://api.hh.ru/vacancies?text={Uri.EscapeDataString(query)}&area=113&per_page=20"
                let! listResp = client.GetAsync(listUrl)
                let! listContent = listResp.Content.ReadAsStringAsync()
                let vacanciesList = deserializeVacanciesList listContent

                // 2 для каждой вакансии полный запрос
                let! fullVacancies =
                    vacanciesList.items
                    |> Seq.map (fun v ->
                        let url = $"https://api.hh.ru/vacancies/{v.id}"
                        task {
                            let! resp = client.GetAsync(url)
                            let! content = resp.Content.ReadAsStringAsync()
                            return deserializeFullVacancy content
                        })
                    |> Task.WhenAll

                // 3 отправляем результат фронтенду
                let jsonResult = serialize fullVacancies
                //return this.Content(jsonResult, "application/json; charset=utf-8") :> IActionResult
                return this.Content(jsonResult, "application/json", System.Text.Encoding.UTF8) :> IActionResult

        }























  

File: Talentsync.CSharp/Controllers/VacanciesController.cs
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Net.Http;
using System.Text;
using Talentsync.CSharp.Model;
using Talentsync.CSharp.Services;

namespace Talentsync.CSharp.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class VacanciesController : ControllerBase
    {
        private readonly IHttpClientFactory _httpClientFactory;
        private readonly IConfiguration _configuration;
        private readonly VacancyDatasetService _datasetService;

        public VacanciesController(
            IHttpClientFactory httpClientFactory,
            IConfiguration configuration,
            VacancyDatasetService datasetService)
        {
            _httpClientFactory = httpClientFactory;
            _configuration = configuration;
            _datasetService = datasetService;
        }

        [HttpGet]
        public IActionResult Get([FromQuery] string? query = "developer")
        {
            var path = Path.Combine(Directory.GetCurrentDirectory(), "Data", "vacancies.csv");

            var data = _datasetService.Load(path);

            Console.WriteLine($"TOTAL: {data.Count}");
            Console.WriteLine($"FIRST: {data.FirstOrDefault()?.Title}");

            if (!string.IsNullOrWhiteSpace(query))
            {
                data = data
                    .Where(v =>
                        v.Title.Contains(query, StringComparison.OrdinalIgnoreCase) ||
                        v.Description.Contains(query, StringComparison.OrdinalIgnoreCase))
                    .ToList();
            }

            return Ok(data);
        }

        [HttpGet("dataset")]
        public IActionResult GetDataset()
        {
            var path = Path.Combine(Directory.GetCurrentDirectory(), "Data", "vacancies.csv");

            var data = _datasetService.Load(path);

            return Ok(data);
        }

        [HttpPost("semantic-search")]
        public async Task<IActionResult> SemanticSearch([FromBody] SearchRequestDto request)
        {
            var client = _httpClientFactory.CreateClient();

            var response = await client.PostAsJsonAsync(
                "http://127.0.0.1:8000/search",
                request
            );

            var json = await response.Content.ReadAsStringAsync();

            return Content(json, "application/json");
        }
    }
}